### PR TITLE
[TS migration] Migrate 'OfflineWithFeedback.js' component to TypeScript

### PR DESCRIPTION
**Pull request by kubabutkiewicz**
### Fixed Issues

$ https://github.com/Expensify/App/issues/25102

### Tests

1. Login to app
2. Open Preferences
3. Go to Offline mode 
4. Start a chat 
5. Send a message
6. Offline indicator should show up

- [x] Verify that no errors appear in the JS console

### QA Steps

- [x] Verify that no errors appear in the JS console

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] If we are not using the full Onyx data that we loaded, I've added the proper selector in order to ensure the component only re-renders when the data it is using changes
        - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/25230417/ccb7a6d7-84b3-4221-8013-6ec14b70165b

</details>

<details>
<summary>Android: mWeb Chrome</summary>

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/25230417/ac73be9a-18dd-4ce2-9898-acaa55fc2091

</details>

<details>
<summary>iOS: mWeb Safari</summary>

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/25230417/a64457f4-1271-4fbb-9c15-a90d0d394567

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/25230417/6e4cccd6-f442-4a22-b6b2-4a473ad87c2f

</details>

### CODE REVIEW

File: src/components/MessagesRow.tsx [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+//     /** Whether we can dismiss the messages */
+//     canDismiss: PropTypes.bool,
+// };
+
+// const defaultProps = {
+//     messages: {},
+//     onClose: () => {},
+//     containerStyles: [],
+//     canDismiss: true,
+// };
```
Comment (VickyStash) [inline src/components/MessagesRow.tsx:36]: Don't forget to remove commented code

File: src/components/MessagesRow.tsx [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+
+type MessagesRowProps = {
+    messages: Record<string, string>;
+    type: 'error' | 'success';
+    onClose: () => void;
+    containerStyles: StyleProp<ViewStyle>;
+    canDismiss: boolean;
```
Comment (VickyStash) [inline src/components/MessagesRow.tsx:43]: Also please, add comments to type props
```suggestion
    onClose?: () => void;
    containerStyles?: StyleProp<ViewStyle>;
    canDismiss?: boolean;
```

File: src/components/MessagesRow.tsx [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+};
+
+function MessagesRow({messages = {}, type, onClose = () => {}, containerStyles, canDismiss = true}: MessagesRowProps) {
+    const styles = useThemeStyles();
+    const {translate} = useLocalize();
```
Comment (VickyStash) [inline src/components/MessagesRow.tsx:48]:
```suggestion
    const {translate} = useLocalize();
    
```

File: src/components/MessagesRow.tsx [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+            <DotIndicatorMessage
+                style={[styles.flex1]}
+                messages={messages}
+                type={type}
+            />
+            {canDismiss && (
+                <Tooltip text={translate('common.close')}>
+                    <PressableWithoutFeedback
+                        onPress={onClose}
+                        style={[styles.touchableButtonImage]}
```
Comment (VickyStash) [inline src/components/MessagesRow.tsx:64]:
```suggestion
                        style={styles.touchableButtonImage}
```

File: src/components/OfflineWithFeedback.tsx [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+    needsOffscreenAlphaCompositing?: boolean;
+
+    /** Whether we can dismiss the error message */
+    canDismissError?: boolean;
+};
+
+/**
+ * This method applies the strikethrough to all the children passed recursively
+ */
+function applyStrikeThrough(children: React.ReactNode) {
```
Comment (VickyStash) [inline src/components/OfflineWithFeedback.tsx:63]: Let's add return type

File: src/components/OfflineWithFeedback.tsx [hunk 6]
```diff
@@ -0,0 +1,136 @@
...
+    return React.Children.map(children, (child) => {
+        if (!React.isValidElement(child)) {
+            return child;
+        }
+        const props: {style: Array<ViewStyle | TextStyle | ImageStyle>; children?: React.ReactNode} = {
```
Comment (VickyStash) [inline src/components/OfflineWithFeedback.tsx:68]: Create separate type for this `{style: Array<ViewStyle | TextStyle | ImageStyle>; children?: React.ReactNode}`

File: src/components/OfflineWithFeedback.tsx [hunk 7]
```diff
@@ -0,0 +1,136 @@
...
+            style: StyleUtils.combineStyles(child.props.style, styles.offlineFeedback.deleted, styles.userSelectNone),
+        };
+        if (child.props.children) {
+            props.children = applyStrikeThrough(child.props.children);
+        }
```
Comment (VickyStash) [inline src/components/OfflineWithFeedback.tsx:73]:
```suggestion
       
         if (child.props.children) {
            props.children = applyStrikeThrough(child.props.children);
        }
        
```

File: src/components/OfflineWithFeedback.tsx [hunk 8]
```diff
@@ -0,0 +1,136 @@
...
+};
+
+/**
+ * This method applies the strikethrough to all the children passed recursively
+ */
+function applyStrikeThrough(children: React.ReactNode) {
+    return React.Children.map(children, (child) => {
+        if (!React.isValidElement(child)) {
+            return child;
+        }
```
Comment (VickyStash) [inline src/components/OfflineWithFeedback.tsx:67]:
```suggestion
        if (!React.isValidElement(child)) {
            return child;
        }
        
```

File: src/components/OfflineWithFeedback.tsx [hunk 9]
```diff
@@ -0,0 +1,136 @@
...
+        return React.cloneElement(child, props);
+    });
+}
+
+function OfflineWithFeedback({
+    pendingAction = null,
+    canDismissError = true,
+    contentContainerStyle = [],
+    errorRowStyles = [],
```
Comment (VickyStash) [inline src/components/OfflineWithFeedback.tsx:82]:
```suggestion
    contentContainerStyle,
    errorRowStyles,
```

File: src/components/OfflineWithFeedback.tsx [hunk 10]
```diff
@@ -0,0 +1,136 @@
...
+    errors = null,
+    needsOffscreenAlphaCompositing = false,
+    onClose = () => {},
+    shouldDisableOpacity = false,
+    shouldDisableStrikeThrough = false,
+    shouldHideOnDelete = true,
+    shouldShowErrorMessages = true,
+    style = [],
```
Comment (VickyStash) [inline src/components/OfflineWithFeedback.tsx:90]:
```suggestion
    style,
```

**Review by VickyStash — CHANGES_REQUESTED** [review]

File: src/components/MessagesRow.tsx (hunk 2 above)
Comment (kubabutkiewicz) [reply src/components/MessagesRow.tsx:43]: I don't know what is happening in my head but I am keep forgetting about that 😅

File: src/components/OfflineWithFeedback.tsx [hunk 11]
```diff
@@ -0,0 +1,140 @@
...
+    shouldDisableStrikeThrough?: boolean;
+
+    /** Whether to apply needsOffscreenAlphaCompositing prop to the children */
+    needsOffscreenAlphaCompositing?: boolean;
+
+    /** Whether we can dismiss the error message */
+    canDismissError?: boolean;
+};
+
+type StrikethroughProps = {style: Array<ViewStyle | TextStyle | ImageStyle>; children?: React.ReactNode};
```
Comment (VickyStash) [inline src/components/OfflineWithFeedback.tsx:60]:
```suggestion
type StrikethroughProps = {style: Array<ViewStyle | TextStyle | ImageStyle>; children?: React.ReactNode};

```

File: src/components/OfflineWithFeedback.tsx [hunk 12]
```diff
@@ -0,0 +1,140 @@
...
+    shouldShowErrorMessages?: boolean;
+
+    /** Whether we should disable opacity */
+    shouldDisableOpacity?: boolean;
+
+    /** A function to run when the X button next to the error is clicked */
+    onClose?: () => void;
+
+    /** The content that needs offline feedback */
+    children: React.ReactNode;
```
[the remark this answers is not in the document]
Comment (VickyStash) [reply src/components/OfflineWithFeedback.tsx:40]: @kubabutkiewicz Why did you skip this one?

[the remark this answers is not in the document]
Comment (kubabutkiewicz) [reply src/components/OfflineWithFeedback.tsx:40]: Oh I added it to another type instead of this 😅

**Review by fabioh8010 — APPROVED** [review]

File: src/types/utils/EmptyObject.ts [hunk 13]
```diff
@@ -0,0 +1,15 @@
+import Falsy from './Falsy';
+
+type EmptyObject = Record<string, never>;
+
+// eslint-disable-next-line rulesdir/no-negated-variables
+function isNotEmptyObject<T extends Record<string, unknown> | Falsy>(arg: T): arg is T {
```
Comment (blazejkustra) [inline src/types/utils/EmptyObject.ts:6]: Please merge main, it should be already there

File: src/components/MessagesRow.tsx [hunk 14]
```diff
@@ -13,43 +9,35 @@ import * as Expensicons from './Icon/Expensicons';
 import PressableWithoutFeedback from './Pressable/PressableWithoutFeedback';
 import Tooltip from './Tooltip';
 
-const propTypes = {
+type MessagesRowProps = {
     /** The messages to display */
-    messages: PropTypes.objectOf(
-        PropTypes.oneOfType([PropTypes.oneOfType([PropTypes.string, PropTypes.object]), PropTypes.arrayOf(PropTypes.oneOfType([PropTypes.string, PropTypes.object]))]),
-    ),
+    messages: Record<string, string>;
```
Comment (blazejkustra) [inline src/components/MessagesRow.tsx:14]: This looks incorrect, check out the type of messages in `DotIndicatorMessage`

File: src/components/MessagesRow.tsx [hunk 15]
```diff
@@ -13,43 +9,35 @@ import * as Expensicons from './Icon/Expensicons';
...
-    canDismiss: true,
-};
-
-function MessagesRow({messages, type, onClose, containerStyles, canDismiss}) {
+function MessagesRow({messages = {}, type, onClose = () => {}, containerStyles, canDismiss = true}: MessagesRowProps) {
     const styles = useThemeStyles();
     const {translate} = useLocalize();
-    if (_.isEmpty(messages)) {
+
+    if (Object.keys(messages).length === 0) {
```
Comment (blazejkustra) [inline src/components/MessagesRow.tsx:33]:
```suggestion
    if (isEmptyObject(messages)) {
```

File: src/components/OfflineWithFeedback.tsx [hunk 16]
```diff
@@ -0,0 +1,139 @@
...
+        if (child.props.children) {
+            props.children = applyStrikeThrough(child.props.children);
+        }
+
+        return React.cloneElement(child, props);
+    });
+}
+
+function OfflineWithFeedback({
+    pendingAction = null,
```
Comment (blazejkustra) [inline src/components/OfflineWithFeedback.tsx:82]: Can we remove the default value and change the type to just `pendingAction: OnyxCommon.PendingAction`?

File: src/components/OfflineWithFeedback.tsx [hunk 17]
```diff
@@ -0,0 +1,139 @@
...
+    canDismissError = true,
+    contentContainerStyle,
+    errorRowStyles,
+    errors = null,
```
Comment (blazejkustra) [inline src/components/OfflineWithFeedback.tsx:86]: Same here ☝️

File: src/components/OfflineWithFeedback.tsx [hunk 18]
```diff
@@ -0,0 +1,139 @@
...
+    shouldHideOnDelete = true,
+    shouldShowErrorMessages = true,
+    style,
+    ...props
+}: OfflineWithFeedbackProps) {
+    const {isOffline} = useNetwork();
+
+    const hasErrors = !isEmptyObject(errors);
+    // Some errors have a null message. This is used to apply opacity only and to avoid showing redundant messages.
+    const errorMessages = lodashOmitBy(errors, (e) => e === null);
```
Comment (blazejkustra) [inline src/components/OfflineWithFeedback.tsx:100]: How about implementation without lodash?

Comment (blazejkustra) [reply src/components/OfflineWithFeedback.tsx:100]: FYI Original implementation used omit, not omitBy:
```ts
    const errorMessages = _.omit(props.errors, (e) => e === null);
````

File: src/components/OfflineWithFeedback.tsx [hunk 19]
```diff
@@ -0,0 +1,139 @@
...
+    shouldDisableOpacity = false,
+    shouldDisableStrikeThrough = false,
+    shouldHideOnDelete = true,
+    shouldShowErrorMessages = true,
+    style,
+    ...props
+}: OfflineWithFeedbackProps) {
+    const {isOffline} = useNetwork();
+
+    const hasErrors = !isEmptyObject(errors);
```
Comment (blazejkustra) [inline src/components/OfflineWithFeedback.tsx:98]:
```suggestion
    const hasErrors = isNotEmptyObject(errors);
```

File: src/components/OfflineWithFeedback.tsx [hunk 20]
```diff
@@ -0,0 +1,139 @@
...
+    errorRowStyles,
+    errors = null,
+    needsOffscreenAlphaCompositing = false,
+    onClose = () => {},
+    shouldDisableOpacity = false,
+    shouldDisableStrikeThrough = false,
+    shouldHideOnDelete = true,
+    shouldShowErrorMessages = true,
+    style,
+    ...props
```
Comment (blazejkustra) [inline src/components/OfflineWithFeedback.tsx:94]:
```suggestion
    ...rest
```

**Review by blazejkustra — CHANGES_REQUESTED** [review]

File: src/components/OfflineWithFeedback.tsx (hunk 18 above)
Comment (kubabutkiewicz) [reply src/components/OfflineWithFeedback.tsx:101]: @blazejkustra lodash omit is used only for omits by keys 
image
for omit by some function `omitBy` need to be used 
image

**Comment by kubabutkiewicz** [comment]
@blazejkustra comments resolved

**Review by BartoszGrajdek — APPROVED** [review]

**Review by blazejkustra — APPROVED** [review]

**Comment by kubabutkiewicz** [comment]
@Ollyws added!

**Review by Ollyws — APPROVED** [review]
LGTM.

**Review by cristipaval — APPROVED** [review]

**Pull request merged**